source/Ship.cpp

```diff
@@ -3669,20 +3683,73 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
 // cues and try to stay with it when it lands or goes into hyperspace.
 void Ship::AddEscort(Ship &ship)
 {
+	TuneForEscort(ship);
 	escorts.push_back(ship.shared_from_this());
 }
 
 
 
 void Ship::RemoveEscort(const Ship &ship)
 {
+	// Reset this value, we will re-scan the list of escorts to set the
+	// escorts velocity based on remaining active escorts.
+	escortsVelocity = -1.;
+
 	auto it = escorts.begin();
-	for( ; it != escorts.end(); ++it)
-		if(it->lock().get() == &ship)
+	while(it != escorts.end())
+	{
+		auto escort = it->lock().get();
+		if(escort == &ship)
+			it = escorts.erase(it);
+		else
 		{
-			escorts.erase(it);
-			return;
+			if(escort)
+				TuneForEscort(*escort);
+			++it;
+		}
+	}
+}
+
+
+
+// (Re)Register escorts, for example because some escort got destroyed.
+void Ship::TuneForEscorts()
+{
+	// Reset this value, we will re-scan the list of escorts to set the
+	// escorts velocity based on remaining active escorts.
+	escortsVelocity = -1.;
+	
+	for(const auto &it: escorts)
+	{
+		auto escort = it.lock().get();
+		if(escort)
+			TuneForEscort(*escort);
+	}
+}
+
+
+
+// Store relevant cached data for the given escort.
+void Ship::TuneForEscort(const Ship &ship)
+{
+	// Cache the maximum speed that the parent should stay below to keep
+	// all escorts together.
+	// We don't cache the speeds of escorts that have 0 velocity to avoid
+	// having the parent get stuck when one of the escorts doesn't have
+	// regular thrust.
+	// We also don't cache the speeds of carried ships, since they are often
+	// docked and the carrier waits for them during docking already.
+	if(!ship.CanBeCarried() && !ship.IsDestroyed() && government == ship.GetGovernment())
+	{
+		double eV = ship.MaxVelocity() * 0.9;
+		if(eV > 0.)
+		{
+			if(escortsVelocity <= 0.)
+				escortsVelocity = eV;
+			else
+				escortsVelocity = min(escortsVelocity, eV);
```

Comment:
Quick observation: Can't you just store the slowest escort? That way, you only need to loop through the whole escort list when that slowest escort dies/gets removed.

Reply:
Good idea! Thanks for the suggestion. I'll add that.

Reply:
I added this to reduce some looping.